In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import pysam

fasta_path = "/home/tnegre/Desktop/matriline_project/matriline/epic_matriline_analysis/GRCm39/bismark_genome/mm39.fa"
fasta = pysam.FastaFile(fasta_path)

# Fetch the full chr1 sequence
chr1_seq = fasta.fetch("chr1")

# Find the first CpG site
idx = chr1_seq.find("CG")
if idx != -1:
    print(f"First CpG at position: {idx + 1}")  # 1-based position
else:
    print("No CpG found on chr1")


First CpG at position: 3050095


In [3]:
cpg_chr1 = "/home/tnegre/Desktop/matriline_project/matriline/epic_matriline_analysis/GRCm39/bismark_genome/CpGs_location_per_region/CpGslocation_chr1_per_region.pkl"
from pyformME import io_informME
df_cpg_chr1 = io_informME.load_pkl(cpg_chr1)

In [13]:
# If df_cpg_chr1 is a DataFrameGroupBy object:
group_keys = df_cpg_chr1.groups.keys()
print(list(group_keys)[0])
df_cpg_chr1.get_group(list(group_keys)[0])

[3050095, 3053095)


,position,next_base,distance_to_next,density,region
0,3050095,G,100.0,0.007,"[3050095, 3053095)"
1,3050195,G,28.0,0.008,"[3050095, 3053095)"
2,3050223,G,39.0,0.008,"[3050095, 3053095)"
3,3050262,G,69.0,0.008,"[3050095, 3053095)"
4,3050331,G,72.0,0.008,"[3050095, 3053095)"
5,3050403,G,153.0,0.008,"[3050095, 3053095)"
6,3050556,G,110.0,0.009,"[3050095, 3053095)"
7,3050666,G,365.0,0.008,"[3050095, 3053095)"
8,3051031,G,169.0,0.004,"[3050095, 3053095)"
9,3051200,G,390.0,0.003,"[3050095, 3053095)"


In [ ]:
bam = io_informME.load_bam("/home/tnegre/Desktop/matriline_project/matriline/epic_matriline_analysis/oocyte_MSUS37_DK/scWGBS/03_pyformME/01_bami/Control/oocyte_3_2_marked.bam")
left = list(group_keys)[0].left
right = list(group_keys)[0].right

reads = bam.fetch("chr1", 3050331, right)


In [26]:
read = list(reads)[0]

In [30]:
print(read)

E00526:575:HJLY3CCX2:5:1107:6167:39423_1:N:0:ACAGTG	1187	chr1	3050283	8	108M	chr1	3050283	108	ACCTATTTAATTCCAAAATCATCATCCATCCCATAATCAACCTCCAAACGCTAACACCATTACATACACTAACCACATTTTACTAATAAAACCCTAATATAACTAAAT	array('B', [32, 32, 32, 37, 37, 41, 37, 41, 41, 37, 41, 41, 41, 41, 41, 41, 41, 41, 41, 37, 32, 37, 22, 41, 41, 41, 41, 41, 37, 41, 37, 32, 41, 41, 41, 41, 41, 41, 41, 32, 41, 41, 37, 41, 41, 41, 41, 41, 41, 41, 41, 37, 41, 41, 41, 41, 41, 37, 41, 41, 41, 41, 37, 37, 41, 41, 37, 32, 41, 41, 41, 41, 41, 37, 41, 37, 41, 41, 37, 41, 41, 41, 41, 41, 41, 41, 41, 41, 37, 41, 41, 41, 41, 41, 41, 41, 41, 41, 37, 41, 37, 41, 41, 41, 41, 32, 41, 41])	[('NM', 17), ('MD', '0G3G10G1G21G12G8G9G9G2G3G0G5G5G2G0T0C1'), ('XM', 'x...x..........x.h.....................x.........Z..x........h.........h.........h..x...hh.....x.....h..x...'), ('XR', 'GA'), ('XG', 'GA'), ('MQ', 8), ('MC', '108M'), ('ms', 3560)]


In [32]:
aligned_pairs

[(0, 3050282),
 (1, 3050283),
 (2, 3050284),
 (3, 3050285),
 (4, 3050286),
 (5, 3050287),
 (6, 3050288),
 (7, 3050289),
 (8, 3050290),
 (9, 3050291),
 (10, 3050292),
 (11, 3050293),
 (12, 3050294),
 (13, 3050295),
 (14, 3050296),
 (15, 3050297),
 (16, 3050298),
 (17, 3050299),
 (18, 3050300),
 (19, 3050301),
 (20, 3050302),
 (21, 3050303),
 (22, 3050304),
 (23, 3050305),
 (24, 3050306),
 (25, 3050307),
 (26, 3050308),
 (27, 3050309),
 (28, 3050310),
 (29, 3050311),
 (30, 3050312),
 (31, 3050313),
 (32, 3050314),
 (33, 3050315),
 (34, 3050316),
 (35, 3050317),
 (36, 3050318),
 (37, 3050319),
 (38, 3050320),
 (39, 3050321),
 (40, 3050322),
 (41, 3050323),
 (42, 3050324),
 (43, 3050325),
 (44, 3050326),
 (45, 3050327),
 (46, 3050328),
 (47, 3050329),
 (48, 3050330),
 (49, 3050331),
 (50, 3050332),
 (51, 3050333),
 (52, 3050334),
 (53, 3050335),
 (54, 3050336),
 (55, 3050337),
 (56, 3050338),
 (57, 3050339),
 (58, 3050340),
 (59, 3050341),
 (60, 3050342),
 (61, 3050343),
 (62, 3050344),
 (

In [35]:
xm_tag = read.get_tag('XM')
aligned_pairs = read.get_aligned_pairs(matches_only=True)
reff_pos = read.get_reference_positions()
print(reff_pos)
print("Read name:", read.query_name)
print("XM tag:", xm_tag)
print("First 10 base positions and XM tag letters:")
for i, (query_pos, ref_pos) in enumerate(aligned_pairs):
    # Note: XM tag should match the number of aligned read bases
    print(f"Base {i}: Genomic position {ref_pos+1}, XM: {xm_tag[i]}")

[3050282, 3050283, 3050284, 3050285, 3050286, 3050287, 3050288, 3050289, 3050290, 3050291, 3050292, 3050293, 3050294, 3050295, 3050296, 3050297, 3050298, 3050299, 3050300, 3050301, 3050302, 3050303, 3050304, 3050305, 3050306, 3050307, 3050308, 3050309, 3050310, 3050311, 3050312, 3050313, 3050314, 3050315, 3050316, 3050317, 3050318, 3050319, 3050320, 3050321, 3050322, 3050323, 3050324, 3050325, 3050326, 3050327, 3050328, 3050329, 3050330, 3050331, 3050332, 3050333, 3050334, 3050335, 3050336, 3050337, 3050338, 3050339, 3050340, 3050341, 3050342, 3050343, 3050344, 3050345, 3050346, 3050347, 3050348, 3050349, 3050350, 3050351, 3050352, 3050353, 3050354, 3050355, 3050356, 3050357, 3050358, 3050359, 3050360, 3050361, 3050362, 3050363, 3050364, 3050365, 3050366, 3050367, 3050368, 3050369, 3050370, 3050371, 3050372, 3050373, 3050374, 3050375, 3050376, 3050377, 3050378, 3050379, 3050380, 3050381, 3050382, 3050383, 3050384, 3050385, 3050386, 3050387, 3050388, 3050389]
Read name: E00526:575:HJLY3

# Test Bam to ME mat

In [ ]:
import pysam
import matplotlib.pyplot as plt

CpG_ref_folder_path = "/Users/tnegre/Desktop/epic_msus/results/test_fasta_cpg"
ref_path = "/Users/tnegre/Desktop/trial_informme/mm10_forWGBS/GRCm38.primary_assembly.genome.fa"



In [19]:

fasta = pysam.FastaFile(ref_path)
def check_bam(bamfile_path,ii):
    bamfile = pysam.AlignmentFile(bamfile_path, "rb")

    # Find one forward and one reverse read
    forward_read = None
    reverse_read = None

    for i, read in enumerate(bamfile.fetch(until_eof=True)):
        if not read.is_unmapped:
            if not read.is_reverse and forward_read is None:
                forward_read = read
            if read.is_reverse and reverse_read is None:
                reverse_read = read
        # if forward_read and reverse_read:
        if i == ii:
            break

    for label, read in [('Forward', forward_read), ('Reverse', reverse_read)]:
        print(f"\n{'='*15} {label} Strand Read {'='*15}")
        print(f"Read name: {read.query_name}")
        print(f"Is reverse: {read.is_reverse}")
        print(f"Read sequence: {read.query_sequence}")
        xm = read.get_tag("XM")
        print(f"XM tag:      {xm}")
        print("Pos | Ref pos | Read base | Ref base | XM")

        # Use the reference name and aligned positions to get ref bases
        aligned_pairs = read.get_aligned_pairs(with_seq=True)
        for i, (read_pos, ref_pos, ref_base) in enumerate(aligned_pairs):
            if read_pos is not None and ref_pos is not None and ref_base is not None:
                read_base = read.query_sequence[read_pos]
                ref_base = ref_base.upper()
                xm_char = xm[read_pos] if read_pos < len(xm) else "."
                print(f"{read_pos:3} | {ref_pos:8} |    {read_base}      |    {ref_base}     | {xm_char}")
        print("="*50)


In [ ]:
dennis_r1_bam = "/Users/tnegre/Desktop/trial_informme/oocyte/oocyte_indexed_bam/PM_3_2-r1_trimmed_bismark_bt2_sm.bam"
dennis_r2_bam = "/Users/tnegre/Desktop/trial_informme/oocyte/oocyte_indexed_bam/PM_3_2-r2_trimmed_bismark_bt2_sm.bam"
chuang_bam = "/Users/tnegre/Desktop/mansuy_server_mnt/mnt/IM/groupMansuy/chuang/MSUS_oocyte_scWGBS_codes/results/01_preprocess/04_bismark_alignment/PM_3_2_sorted.bam"
bam_path_sperm = "/Users/tnegre/Desktop/trial_informme/sperm/bam_example/sub_F1_2_S20.bam"

In [ ]:
check_bam(dennis_r2_bam)

In [ ]:
check_bam(dennis_r1_bam)

In [ ]:
check_bam(chuang_bam)

In [26]:
check_bam(bam_path_sperm, ii=13)


=============== Forward Strand Read ===============
Read name: LH00289:227:22VFG2LT3:3:1212:40425:29455_1:N:0:CTTCACTG+GTCCTAAG
Is reverse: False
Read sequence: TAAAGTAAGTAAGTATTTTTTATTTGGGATGTTTTTGTGGTGGGTTTTGTTTTTATTTTATAATGTTAGGTGATTGGTTAGTTTTATGATTAATGAAGGATGATAGAATTTTTTAGGGAAGTTATTATATAAGTAAT
XM tag:      .....h............h.h.......................h.......h...h...h.....x..........h...hh......h....................h..........h...h.......h..h
Pos | Ref pos | Read base | Ref base | XM
  0 |  3999872 |    T      |    T     | .
  1 |  3999873 |    A      |    A     | .
  2 |  3999874 |    A      |    A     | .
  3 |  3999875 |    A      |    A     | .
  4 |  3999876 |    G      |    G     | .
  5 |  3999877 |    T      |    C     | h
  6 |  3999878 |    A      |    A     | .
  7 |  3999879 |    A      |    A     | .
  8 |  3999880 |    G      |    G     | .
  9 |  3999881 |    T      |    T     | .
 10 |  3999882 |    A      |    A     | .
 11 |  3999883 |    A      |    A     | .
 12

In [ ]:
fasta_path = "/Users/tnegre/Desktop/trial_informme/mm10_forWGBS/GRCm38.primary_assembly.genome.fa"

fa = pysam.FastaFile(fasta_path)  # Use pysam's FastaFile

with pysam.AlignmentFile(bam_path_sperm, "rb") as bam:
    for i, read in enumerate(bam.fetch(until_eof=True)):
        if i > 15:
            break
        read_name = read.query_name
        sequence = read.query_sequence
        strand = "reverse" if read.is_reverse else "forward"
        try:
            xm_tag = read.get_tag("XM")
        except KeyError:
            xm_tag = "NA"
        
        # Only fetch if read is mapped
        if not read.is_unmapped:
            ref_name = read.reference_name
            ref_start = read.reference_start  # 0-based, inclusive
            ref_end = read.reference_end      # 0-based, exclusive
            ref_seq = fa.fetch(ref_name, ref_start, ref_end)
        else:
            ref_seq = "unmapped"

        print(f"{read_name}\n{strand}\n{sequence}\n{xm_tag}\n{ref_seq}\n")

LH00289:227:22VFG2LT3:3:1212:40425:29455_1:N:0:CTTCACTG+GTCCTAAG
forward
TAAAGTAAGTAAGTATTTTTTATTTGGGATGTTTTTGTGGTGGGTTTTGTTTTTATTTTATAATGTTAGGTGATTGGTTAGTTTTATGATTAATGAAGGATGATAGAATTTTTTAGGGAAGTTATTATATAAGTAAT
.....h............h.h.......................h.......h...h...h.....x..........h...hh......h....................h..........h...h.......h..h
TAAAGCAAGTAAGTATTTCTCATTTGGGATGTTTTTGTGGTGGGCTTTGTTTCTATCTTACAATGTCAGGTGATTGGCTAGCCTTATGACTAATGAAGGATGATAGAATTCTTTAGGGAAGCTATCATATAAGCAAC

LH00289:227:22VFG2LT3:3:1213:38345:7297_1:N:0:CTTCACTG+GTCCTAAG
forward
TAAAGTAAGTAAGTATTTTTTATTTGGGATGTTTTTGTGGTGGGTTTTGTTTTTATTTTATAATGTTAGGTGATTGGTTAGTTTTATGATTAATGAAGGATGATAGAATTTTTTAGGGAAGTTATTATATAAGTAAT
.....h............h.h.......................h.......h...h...h.....x..........h...hh......h....................h..........h...h.......h..h
TAAAGCAAGTAAGTATTTCTCATTTGGGATGTTTTTGTGGTGGGCTTTGTTTCTATCTTACAATGTCAGGTGATTGGCTAGCCTTATGACTAATGAAGGATGATAGAATTCTTTAGGGAAGCTATCATATAAGCAAC

LH00289:227:22VFG2LT3:3:2

In [85]:
import pysam
import utils_informME

# Replace 'yourfile.bam' with your actual BAM file name
bamfile = pysam.AlignmentFile(dennis_r1_bam, 'rb')
chrom = 'chr1'  # or 'chr1' if that's how your file labels it

chrom = utils_informME.chr2refseq(chrom)
start = 5000000
end = 5003000

for read in bamfile.fetch(chrom, start, end):
    ref_pos = read.get_reference_positions()
    read_seq = read.query_sequence
    XM = read.get_tag('XM')
    xr_tag = read.get_tag('XR')
    xg_tag = read.get_tag('XG')

    print(read)
    print(read.flag)
    print("Origin read:", read_seq)
    print("           :", XM)
    print("           :", xr_tag)
    print("           :", xg_tag)
    print("Proper pair:", read.is_proper_pair)
    print("Ref start  :", read.reference_start)
    print("Ref end    :", read.reference_end)
    print("Duplicate  :", read.is_duplicate)
    print("Reverse    :", read.is_reverse)
    print(read.reference_name)
    print("-----------")
bamfile.close()


E00526:575:HJLY3CCX2:5:1216:13951:25200_1:N:0:ACAGTG	16	NC_000067.6	5000230	42	77M	*	0	0	AAAAAATTCTTAAAAAAAAAAATTTATATAACTAAAAAACATTTTTAAAATATTCAACATCCTTAATCATCAAAAAA	array('B', [37, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 37, 41, 41, 41, 41, 41, 37, 41, 41, 41, 37, 41, 41, 41, 41, 41, 41, 41, 32, 37, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 32, 41, 41, 37, 37, 41, 41, 37, 41, 32, 41, 37, 37, 12, 37, 41, 27, 37, 41, 41, 41, 41])	[('NM', 14), ('MD', '1G1G11G1G0G10G0G2G1G15G13G6G0G0G2'), ('XM', '.x.h...........h.hh..........hh..x.h...............h.............h......xhh..'), ('XR', 'GA'), ('XG', 'GA')]
16
Origin read: AAAAAATTCTTAAAAAAAAAAATTTATATAACTAAAAAACATTTTTAAAATATTCAACATCCTTAATCATCAAAAAA
           : .x.h...........h.hh..........hh..x.h...............h.............h......xhh..
           : GA
           : GA
Proper pair: False
Ref start  : 5000229
Ref end    : 5000306
Duplicate  : False
Reverse    : True
NC_000067.6
----------

In [81]:
from collections import defaultdict

# Replace 'yourfile.bam' with your actual BAM file name
bamfile = pysam.AlignmentFile(dennis_r1_bam, 'rb')
chrom = 'chr1'  # or 'chr1' if that's how your file labels it

chrom = utils_informME.chr2refseq(chrom)
start = 5015000
end = 5025000

reads = list(bamfile.fetch(chrom, start, end))
dupe_groups = defaultdict(list)
for i, read in enumerate(reads):

    ref_pos = read.get_reference_positions()
    read_seq = read.query_sequence
    XM = read.get_tag('XM')
    xr_tag = read.get_tag('XR')
    xg_tag = read.get_tag('XG')
    print(i)
    print(read)
    print(read.flag)
    print("Origin read:", read_seq)
    print("           :", XM)
    print("           :", xr_tag)
    print("           :", xg_tag)
    print("Proper pair:", read.is_proper_pair)
    print("Ref start  :", read.reference_start)
    print("Ref end    :", read.reference_end)
    print("Duplicate  :", read.is_duplicate)
    print("Reverse    :", read.is_reverse)
    # For paired-end, only process one read per pair (commonly read1)
    if read.is_paired and read.is_proper_pair and read.is_read1:
        # Template start = min(read start, mate start)
        # Template end = max(read end, mate end)
        t_start = min(read.reference_start, read.next_reference_start)
        print(read.reference_start, read.next_reference_start)
        # read.reference_end gives the end of this read
        # mate end estimated as next_reference_start + read.query_length (may not be precise for split reads)
        t_end = max(read.reference_end, read.next_reference_start + read.query_length)
        print(read.reference_end, read.next_reference_start + read.query_length)
        key = (read.reference_name, t_start, t_end, read.is_reverse)
    else:
        # Single-end or unmapped/other, use read start, strand
        key = (read.reference_name, read.reference_start, read.reference_end, read.is_reverse)
        print(read.reference_end, read.next_reference_start, read.query_length)

    dupe_groups[key].append(read)
    print("-----------")

unique_reads = []
for group in dupe_groups.values():
    best_read = max(group, key=lambda r: r.mapping_quality)
    unique_reads.append(best_read)


0
E00526:575:HJLY3CCX2:5:2213:28107:12103_1:N:0:ACAGTG	0	NC_000067.6	5015932	42	144M	*	0	0	TTTAGTTGTTTTTGTGAGGGTGGTTTTTTTAAGTTGTTTTTTATATTTTTCTTATATATAGTGAGTATGGTTTTATAGAAAGATAGATTTGAATATTTAAGTAAGATATATAGTAAGATTTTAAGAATTGGAATGTGATATTTT	array('B', [37, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 37, 41, 37, 41, 41, 37, 37, 41, 37, 41, 41, 41, 41, 41, 37, 37, 41, 41, 41, 41, 37, 27, 27, 41, 41, 41, 41])	[('NM', 28), ('MD', '0C1C2C4C0C11C1C0C6C3C0C1C0C1C2C2C2C5C13C0C13C9C12C16C10C2C0C0C0'), ('XM', 'h.x..x....hx...........h.hh......x...hh.hh.h..h..hH.h.....x.............hh.....

In [80]:
samp_1 = reads[3].query_qualities
samp_2 = reads[4].query_qualities


In [75]:
import numpy as np

In [76]:
np.sum(samp_1)

np.uint64(2333)

In [77]:
np.sum(samp_2)

np.uint64(2333)

In [82]:
bam_full_sperm_pat = "/Users/tnegre/Desktop/trial_informme/sperm/bam_example/MSUS23-F1-2_S20_R1_val_1_bismark_bt2_pe_sm.bam"

In [84]:
# Replace 'yourfile.bam' with your actual BAM file name
bamfile = pysam.AlignmentFile(bam_full_sperm_pat, 'rb')
chrom = 'chr1'  # or 'chr1' if that's how your file labels it

start = 5000000
end = 5003000

for read in bamfile.fetch(chrom, start, end):
    ref_pos = read.get_reference_positions()
    read_seq = read.query_sequence
    XM = read.get_tag('XM')
    xr_tag = read.get_tag('XR')
    xg_tag = read.get_tag('XG')

    print(read)
    print(read.flag)
    print("Origin read:", read_seq)
    print("           :", XM)
    print(read.is_read1)
    print("xr         :", xr_tag)
    print("xg         :", xg_tag)
    print("Proper pair:", read.is_proper_pair)
    print("Ref start  :", read.reference_start)
    print("Ref end    :", read.reference_end)
    print("Duplicate  :", read.is_duplicate)
    print("Reverse    :", read.is_reverse)
    print("-----------")
bamfile.close()


LH00289:227:22VFG2LT3:5:2193:51373:2057_1:N:0:CTTCACTG+GTCATAAG	163	chr1	4999873	42	151M	chr1	4999902	177	TAATAAATATACTTAAAAAACTTCTATATAACAAAAAACACCACCATTTAAACAAAATAACAAACTACAACAAAAAACAAATTTATACAAACTATACATCTAATAAAAAACTAATCTCCAAAATATATAAAAAACTCAAAAAAATAAATAT	array('B', [40, 40, 24, 40, 24, 24, 40, 40, 40, 40, 24, 40, 40, 40, 40, 40, 40, 40, 24, 40, 24, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 24, 40, 40, 12, 24, 40, 40, 12, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 12, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 24, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 24, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 24, 40, 24, 40, 40, 40, 40, 40, 40, 40, 24, 40, 40, 40, 40, 40, 40, 40, 40, 40, 12, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40])	[('NM', 32), ('MD', '1G6G1G3G1G1G0G5G3G0G4G0G12G8G0G2G0G2G2G3G0G0G0G3G3G16G3G1G0G0G21G14G4'), ('XM', '.h......h.h...h.h.hh...

E00526:575:HJLY3CCX2:5:1216:25205:28945_1:N:0:ACAGTG	1040	NC_000067.6	5002531	42	57M	*	0	0	TCATCAAAAAAAATAAAAACAAATACAAAAACTCACAACCAAACAACACACAAAAAA	array('B', [37, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41, 41])	[('NM', 13), ('MD', '6G1G2G2G1G1G8G1G7G3G10G1G1G0'), ('XM', '......x.h..h..h.h.h........x.h.......x...x..........x.h.h'), ('XR', 'GA'), ('XG', 'GA')]


In [93]:
import pysam
from collections import defaultdict

def count_reverse_by_xr_xg(bamfile_path):
    """
    Count how many reads in a region are reverse, grouped by (XR, XG) tag values.

    Args:
        bamfile_path (str): Path to BAM file.
        region (str): Region to fetch, e.g., 'chr1:1000-2000'
    """
    bamfile = pysam.AlignmentFile(bamfile_path, "rb")
    # Dict to hold: {(XR, XG): [total_count, reverse_count]}
    counter = defaultdict(lambda: [0, 0])
    chrom = 'chr1'
    chrom = utils_informME.chr2refseq(chrom)
    start = 5015000
    end = 5105000

    reads = list(bamfile.fetch(chrom, start, end))
    for read in reads:
        try:
            xr = read.get_tag('XR')
            xg = read.get_tag('XG')
        except KeyError:
            continue  # Skip reads missing these tags

        counter[(xr, xg)][0] += 1
        if read.is_reverse:
            counter[(xr, xg)][1] += 1

    # Print results
    print("XR\tXG\tTotal\tReverse")
    for (xr, xg), (total, reverse) in counter.items():
        print(f"{xr}\t{xg}\t{total}\t{reverse}")

# Example usage:
# count_reverse_by_xr_xg("example.bam", "chr1:1000-2000")


In [92]:
count_reverse_by_xr_xg(bamfile_path=bam_full_sperm_pat)

XR	XG	Total	Reverse
GA	GA	3421	0
CT	GA	3422	3422
CT	CT	3151	0
GA	CT	3151	3151


In [94]:
count_reverse_by_xr_xg(bamfile_path=dennis_r1_bam)

XR	XG	Total	Reverse
CT	CT	64	0
GA	GA	85	85
CT	GA	81	81
GA	CT	75	0


In [95]:
count_reverse_by_xr_xg(bamfile_path=dennis_r2_bam)

XR	XG	Total	Reverse
GA	CT	48	0
CT	CT	63	0
GA	GA	64	64
CT	GA	65	65
